File: projects/klickety/js/klickety.js
/**
 * Klickety - Jeu de puzzle où vous éliminez des groupes de blocs de même couleur
 * Conversion JavaScript du jeu Python original
 */

class Klickety {
    constructor(canvasId, blocCountCanvasId) {
        // Initialisation des structures de données
        this.gridRows = 16;      // nombre de lignes du plateau
        this.gridCols = 10;      // nombre de colonnes du plateau
        this.cellSize = 32;      // taille d'une case en pixels
        this.boardHeight = this.cellSize * this.gridRows;
        this.boardWidth = this.cellSize * this.gridCols;
        this.grid = [];
        this.colors = ["red", "blue", "green", "yellow", "magenta"];
        this.blocCount = 0;
        
        // Initialisation des éléments graphiques
        this.canvas = document.getElementById(canvasId);
        this.ctx = this.canvas.getContext('2d');
        this.canvas.width = this.boardWidth;
        this.canvas.height = this.boardHeight;
        
        this.blocCountCanvas = document.getElementById(blocCountCanvasId);
        this.blocCountCtx = this.blocCountCanvas.getContext('2d');
        this.blocCountCanvas.width = this.boardWidth;
        this.blocCountCanvas.height = 32;
        
        // Ajout des écouteurs d'événements
        this.canvas.addEventListener('click', this.handleClick.bind(this));
        
        // Initialisation du jeu
        this.initGame();
    }
    
    initGame() {
        // Réinitialisation du jeu
        this.initGrid();
        this.refreshBlocCount();
    }
    
    initGrid() {
        // Initialise la grille avec des blocs de couleurs aléatoires
        this.grid = [];
        for (let i = 0; i < this.gridRows; i++) {
            let row = [];
            for (let j = 0; j < this.gridCols; j++) {
                row.push(this.colors[Math.floor(Math.random() * this.colors.length)]);
            }
            this.grid.push(row);
        }
        
        this.refreshBoard();
    }
    
    refreshBoard() {
        // Redessine le plateau de jeu
        this.ctx.clearRect(0, 0, this.boardWidth, this.boardHeight);
        
        // Dessiner les blocs
        const backgroundColor = "black";
        
        for (let i = 0; i < this.gridRows; i++) {
            for (let j = 0; j < this.gridCols; j++) {
                const cell = this.grid[i][j];
                
                if (cell !== null) {
                    // Dessiner un bloc coloré
                    this.ctx.fillStyle = cell;
                    this.ctx.fillRect(j * this.cellSize, i * this.cellSize, this.cellSize, this.cellSize);
                    this.ctx.strokeStyle = cell;
                    this.ctx.strokeRect(j * this.cellSize, i * this.cellSize, this.cellSize, this.cellSize);
                } else {
                    // Dessiner un espace vide
                    this.ctx.fillStyle = backgroundColor;
                    this.ctx.fillRect(j * this.cellSize, i * this.cellSize, this.cellSize, this.cellSize);
                    this.ctx.strokeStyle = backgroundColor;
                    this.ctx.strokeRect(j * this.cellSize, i * this.cellSize, this.cellSize, this.cellSize);
                }
            }
        }
        
        // Tracer le contour des pièces
        // 1) Séparations entre deux pièces adjacentes de couleurs différentes dans la même colonne
        for (let j = 0; j < this.gridCols; j++) {
            for (let i = 1; i < this.gridRows; i++) {
                if (this.grid[i-1][j] !== this.grid[i][j] && 
                    this.grid[i-1][j] !== null && this.grid[i][j] !== null) {
                    this.ctx.fillStyle = backgroundColor;
                    this.ctx.fillRect(j * this.cellSize, i * this.cellSize, this.cellSize, 1);
                }
            }
        }
        
        // 2) Séparations entre deux pièces adjacentes de couleurs différentes dans la même ligne
        for (let i = 0; i < this.gridRows; i++) {
            for (let j = 1; j < this.gridCols; j++) {
                if (this.grid[i][j-1] !== this.grid[i][j] && 
                    this.grid[i][j-1] !== null && this.grid[i][j] !== null) {
                    this.ctx.fillStyle = backgroundColor;
                    this.ctx.fillRect(j * this.cellSize, i * this.cellSize, 1, this.cellSize);
                }
            }
        }
    }
    
    refreshBlocCount(piece = null) {
        // Rafraîchit l'affichage du nombre de blocs restants
        this.blocCountCtx.clearRect(0, 0, this.blocCountCanvas.width, this.blocCountCanvas.height);
        
        if (piece === null) {
            // Appel initial, tous les blocs sont encore présents
            this.blocCount = this.gridRows * this.gridCols;
        } else {
            // Soustraire du nombre de blocs celui de la pièce retirée
            this.blocCount -= piece.size;
        }
        
        this.blocCountCtx.fillStyle = "black";
        this.blocCountCtx.font = "14px Arial";
        this.blocCountCtx.textAlign = "center";
        this.blocCountCtx.fillText("Blocs restants: " + this.blocCount, this.boardWidth / 2, this.cellSize / 2 + 5);
    }
    
    handleClick(event) {
        // Gère les clics sur le plateau
        const rect = this.canvas.getBoundingClientRect();
        const x = event.clientX - rect.left;
        const y = event.clientY - rect.top;
        
        const col = Math.floor(x / this.cellSize);
        const row = Math.floor(y / this.cellSize);
        
        if (row >= 0 && row < this.gridRows && col >= 0 && col < this.gridCols) {
            if (this.grid[row][col] !== null) {
                const piece = this.detectPiece(row, col);
                
                if (piece.size > 1) {  // Si la pièce est valide (plus d'un bloc), on la retire
                    // Retirer la pièce en mettant ses cases à null
                    for (let coords of piece) {
                        const [r, c] = coords.split(',').map(Number);
                        this.grid[r][c] = null;
                    }
                    
                    // Faire descendre les blocs situés au-dessus de la pièce
                    this.updateGrid(piece);
                    
                    // Tasser le restant du plateau en supprimant les colonnes vides
                    this.removeEmptyColumns();
                    
                    // Rafraîchir le plateau pour répercuter les modifications
                    this.refreshBoard();
                    this.refreshBlocCount(piece);
                    
                    if (this.isGameOver()) {
                        this.ctx.fillStyle = "red";
                        this.ctx.font = "bold 16px Courier";
                        this.ctx.textAlign = "center";
                        this.ctx.fillText("LA PARTIE EST TERMINÉE", this.boardWidth / 2, this.cellSize / 2 + 5);
                    }
                }
            }
        }
    }
    
    detectPiece(row, col) {
        // Détecte tous les blocs connectés de même couleur
        const piece = new Set();
        const color = this.grid[row][col];
        
        const explore = (r, c) => {
            const key = `${r},${c}`;
            if (piece.has(key) || this.grid[r][c] !== color) return;
            
            piece.add(key);
            
            // Directions: haut, bas, gauche, droite
            const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
            
            for (let [dr, dc] of directions) {
                const nr = r + dr;
                const nc = c + dc;
                
                if (nr >= 0 && nr < this.gridRows && nc >= 0 && nc < this.gridCols && 
                    this.grid[nr][nc] === color) {
                    explore(nr, nc);
                }
            }
        };
        
        explore(row, col);
        return piece;
    }
    
    updateGrid(piece) {
        // Fait descendre les blocs après suppression d'une pièce
        // Parcourir chaque colonne
        for (let j = 0; j < this.gridCols; j++) {
            // Extraire les blocs non vides de la colonne
            let column = [];
            for (let i = 0; i < this.gridRows; i++) {
                if (this.grid[i][j] !== null) {
                    column.push(this.grid[i][j]);
                }
            }
            
            // Compléter avec null pour atteindre la hauteur du plateau
            column = Array(this.gridRows - column.length).fill(null).concat(column);
            
            // Réinsérer la colonne mise à jour dans la matrice
            for (let i = 0; i < this.gridRows; i++) {
                this.grid[i][j] = column[i];
            }
        }
    }
    
    removeEmptyColumns() {
        // Supprime les colonnes vides et décale les autres
        // Identifier les colonnes non vides
        let nonEmptyColumns = [];
        for (let j = 0; j < this.gridCols; j++) {
            let isEmpty = true;
            for (let i = 0; i < this.gridRows; i++) {
                if (this.grid[i][j] !== null) {
                    isEmpty = false;
                    break;
                }
            }
            if (!isEmpty) {
                nonEmptyColumns.push(j);
            }
        }
        
        // Créer une nouvelle grille avec les colonnes non vides à gauche
        let newGrid = Array(this.gridRows).fill().map(() => Array(this.gridCols).fill(null));
        
        for (let newCol = 0; newCol < nonEmptyColumns.length; newCol++) {
            let oldCol = nonEmptyColumns[newCol];
            for (let i = 0; i < this.gridRows; i++) {
                newGrid[i][newCol] = this.grid[i][oldCol];
            }
        }
        
        this.grid = newGrid;
    }
    
    isGameOver() {
        // Vérifie si la partie est terminée
        // La partie est terminée si le plateau est vide
        let isEmpty = true;
        for (let i = 0; i < this.gridRows; i++) {
            for (let j = 0; j < this.gridCols; j++) {
                if (this.grid[i][j] !== null) {
                    isEmpty = false;
                    break;
                }
            }
            if (!isEmpty) break;
        }
        
        if (isEmpty) return true;
        
        // Ou si toutes les pièces restantes sont de taille 1
        for (let i = 0; i < this.gridRows; i++) {
            for (let j = 0; j < this.gridCols; j++) {
                if (this.grid[i][j] !== null) {
                    const piece = this.detectPiece(i, j);
                    if (piece.size > 1) {
                        return false;
                    }
                }
            }
        }
        
        return true;
    }
    
    reset() {
        // Réinitialise le jeu
        this.initGame();
    }
}

// Initialisation du jeu lorsque la page est chargée
document.addEventListener('DOMContentLoaded', function() {
    const resetButton = document.getElementById('reset-button');
    const game = new Klickety('game-canvas', 'bloc-count-canvas');
    
    if (resetButton) {
        resetButton.addEventListener('click', function() {
            game.reset();
        });
    }
});

File: projects/python-files/klickety.py
# Imports ---------------------------------------------------------------------
from tkinter import Tk, Frame, LEFT, RIGHT, Button, BOTH, Canvas, TOP, \
                    BOTTOM, ALL
import tkinter.font
import random

COULEURS = ["red", "blue", "green", "yellow", "magenta"]

def initialiser_plateau(m, n):
    """Renvoie un plateau m x n aléatoire de blocs de couleurs."""
    return [[random.choice(COULEURS) for _ in range(n)] for _ in range(m)]

def detecter_piece(M, i, j, piece):
    """Remplit l'ensemble piece, initialement vide, à l'aide des coordonnées
    des entrées de M appartenant à la même pièce que M[i][j]."""
    if (i, j) in piece or M[i][j] is None:
        return

    piece.add((i, j))
    couleur = M[i][j]
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    for di, dj in directions:
        ni, nj = i + di, j + dj
        if 0 <= ni < len(M) and 0 <= nj < len(M[0]) and M[ni][nj] == couleur:
            detecter_piece(M, ni, nj, piece)

def mettre_a_jour(M, piece):
    """Modifie M de manière à ce que les trous liés à la suppression de la
    pièce donnée fassent chuter les autres blocs. Les coordonnées renseignées
    par piece renseignent des cases déjà à None dans M."""
    # Parcourir chaque colonne
    for j in range(len(M[0])):
        # Extraire les blocs non vides de la colonne
        colonne = [M[i][j] for i in range(len(M)) if M[i][j] is not None]
        # Compléter avec None pour atteindre la hauteur du plateau
        colonne = [None] * (len(M) - len(colonne)) + colonne
        # Réinsérer la colonne mise à jour dans la matrice
        for i in range(len(M)):
            M[i][j] = colonne[i]

def eliminer_colonnes_vides(M):
    """Effectue les décalages nécessaires à la suppression des colonnes
    vides."""
    # Extraire les colonnes non vides
    colonnes_non_vides = [
        [M[i][j] for i in range(len(M))] for j in range(len(M[0]))
        if any(M[i][j] is not None for i in range(len(M)))
    ]
    # Compléter avec des colonnes vides (None) pour conserver la largeur
    colonnes_vides = [[None for _ in range(len(M))]] * (len(M[0]) - len(colonnes_non_vides))
    colonnes_reorganisees = colonnes_non_vides + colonnes_vides

    # Réinsérer les colonnes dans la matrice
    for j in range(len(M[0])):
        for i in range(len(M)):
            M[i][j] = colonnes_reorganisees[j][i]

def partie_finie(M):
    """Renvoie True si la partie est finie, c'est-à-dire si le plateau est vide
    ou si les seules pièces restantes sont de taille 1, et False sinon"""
    # Vérifier si le plateau est vide
    if all(all(cell is None for cell in row) for row in M):
        return True

    # Vérifier si toutes les pièces restantes sont de taille 1
    for i in range(len(M)):
        for j in range(len(M[0])):
            if M[i][j] is not None:
                piece = set()
                detecter_piece(M, i, j, piece)
                if len(piece) > 1:
                    return False

    return True

# =============================================================================
# PARTIE A NE PAS MODIFIER ====================================================
# =============================================================================


class KlicketyGUI:
    """Interface pour le jeu Klickety."""
    def __init__(self):
        # initialisation des structures de données ----------------------------
        self.dim_plateau = (16,                 # nombre de lignes du plateau
                            10)                 # nombre de colonnes du plateau
        self.cote_case = 32          # la longueur du côté d'un bloc à dessiner
        self.hauteur_plateau = self.cote_case * self.dim_plateau[0]
        self.largeur_plateau = self.cote_case * self.dim_plateau[1]
        self.plateau = []

        # initialisation des éléments graphiques ------------------------------
        self.window = Tk()                              # la fenêtre principale
        self.window.resizable(0, 0)           # empêcher les redimensionnements
        self.partie_haut = Frame(
            self.window, width=self.largeur_plateau,
            height=self.hauteur_plateau
        )
        self.partie_haut.pack(side=TOP)
        self.partie_bas = Frame(self.window)
        self.partie_bas.pack(side=BOTTOM)

        # le canevas affichant le plateau de jeu
        self.plateau_affiche = Canvas(self.partie_haut,
                                      width=self.largeur_plateau,
                                      height=self.hauteur_plateau)
        self.plateau_affiche.pack()
        self.plateau_affiche.bind('<ButtonPress-1>', self.clic_plateau)

        # le bouton "Réinitialiser"
        self.btn = Button(self.partie_bas, text='Réinitialiser',
                          command=self.reinitialiser_jeu)
        self.btn.pack(fill=BOTH)

        # affichage du nombre de blocs restants
        self.nb_blocs = 0
        self.nb_blocs_affiche = Canvas(self.partie_bas,
                                       width=self.largeur_plateau, height=32)
        self.nb_blocs_affiche.pack(fill=BOTH)

        self.reinitialiser_jeu()

        self.window.title('Klickety')
        self.window.mainloop()

    def rafraichir_nombre_blocs(self, piece=None):
        """Rafraîchit l'affichage du nombre de blocs restants, sur base de la
        pièce que l'on vient de retirer."""
        self.nb_blocs_affiche.delete(ALL)
        if piece is None:  # appel initial, tous les blocs sont encore présents
            self.nb_blocs = self.dim_plateau[0] * self.dim_plateau[1]

        else:  # soustraire du nombre de blocs celui de la pièce retirée
            self.nb_blocs -= len(piece)

        self.nb_blocs_affiche.create_text(
            self.largeur_plateau // 2, self.cote_case // 2,
            text="Blocs restants: " + str(self.nb_blocs), fill="black"
        )

    def rafraichir_plateau(self):
        """Redessine le plateau de jeu à afficher."""
        # tracer les blocs
        self.plateau_affiche.delete(ALL)
        couleur_fond = "black"
        for i in range(self.dim_plateau[1]):                    # par défaut 10
            for j in range(self.dim_plateau[0]):                # par défaut 16
                # remarque: le canevas de tkinter interprète (i, j)
                # géométriquement (au lieu de (ligne, colonne)), d'où
                # l'inversion de coordonnées dans la ligne ci-dessous
                case = self.plateau[j][i]
                if case is not None:  # afficher le pion
                    self.plateau_affiche.create_rectangle(
                        i * self.cote_case, j * self.cote_case,
                        (i + 1) * self.cote_case, (j + 1) * self.cote_case,
                        outline=case, fill=case
                    )
                else:
                    self.plateau_affiche.create_rectangle(
                        i * self.cote_case, j * self.cote_case,
                        (i + 1) * self.cote_case, (j + 1) * self.cote_case,
                        outline=couleur_fond, fill=couleur_fond
                    )

        # tracer le contour des pièces
        # 1) tracer les séparations entre deux pièces adjacentes de
        # couleurs différentes dans la même colonne
        for i in range(0, self.dim_plateau[1]):                 # par défaut 10
            for j in range(1, self.dim_plateau[0]):             # par défaut 16
                if self.plateau[j - 1][i] != self.plateau[j][i]:
                    self.plateau_affiche.create_rectangle(
                        i * self.cote_case, j * self.cote_case,
                        (i + 1) * self.cote_case, j * self.cote_case,
                        outline=couleur_fond, fill=couleur_fond, width=1
                    )

        # 2) tracer les séparations entre deux pièces adjacentes de
        # couleurs différentes dans la même ligne
        for i in range(1, self.dim_plateau[1]):                 # par défaut 10
            for j in range(0, self.dim_plateau[0]):             # par défaut 16
                if self.plateau[j][i - 1] != self.plateau[j][i]:
                    self.plateau_affiche.create_rectangle(
                        i * self.cote_case, j * self.cote_case,
                        i * self.cote_case, (j + 1) * self.cote_case,
                        outline=couleur_fond, fill=couleur_fond, width=1
                    )

    def clic_plateau(self, event):
        """Récupère les coordonnées de la case sélectionnée, et joue le coup
        correspondant s'il est permis."""
        # remarque: le canevas de tkinter interprète (i, j) géométriquement
        # (au lieu de (ligne, colonne)), d'où l'inversion de coordonnées dans
        # la ligne ci-dessous
        (j, i) = (event.x // self.cote_case, event.y // self.cote_case)

        if self.plateau[i][j] is not None:
            piece = set()
            detecter_piece(self.plateau, i, j, piece)

            if len(piece) > 1:  # si la pièce est valide, on la retire
                # retirer la piece en mettant ses cases à None
                for (p, q) in piece:
                    self.plateau[p][q] = None

                # faire descendre les blocs situés au-dessus de la pièce
                mettre_a_jour(self.plateau, piece)

                # tasser le restant du plateau en supprimant les colonnes vides
                eliminer_colonnes_vides(self.plateau)

                # rafraîchir le plateau pour répercuter les modifications
                self.rafraichir_plateau()

                self.rafraichir_nombre_blocs(piece)
                if partie_finie(self.plateau):
                    self.plateau_affiche.create_text(
                        int(self.plateau_affiche.cget("width")) // 2,
                        self.cote_case // 2,
                        text="LA PARTIE EST TERMINÉE",
                        font=tkinter.font.Font(
                            family="Courier", size=12, weight=tkinter.font.BOLD
                        ),
                        fill="red"
                    )

    def reinitialiser_jeu(self):
        """Réinitialise le plateau de jeu et les scores."""
        self.reinitialiser_plateau()
        self.rafraichir_nombre_blocs()

    def reinitialiser_plateau(self):
        """Réinitialise le plateau de jeu."""
        # réinitialiser la matrice
        self.plateau = initialiser_plateau(*self.dim_plateau)

        # réinitialiser l'affichage
        self.plateau_affiche.delete(ALL)

        if self.plateau is not None:
            self.rafraichir_plateau()


if __name__ == "__main__":
    KlicketyGUI()
